File: src/controller.py
import numpy as np
from include import constants as cons
from include.mixer import get_rpm
from src.gravity import gravity
from src.guidance import guidance

# This is the controller responsible to calculate the required RPMs at any point of time
def contorller_output(t,state,extra_paras=None):

    extra=extra_paras[0:3]
    dt=extra_paras[2]
    motors=extra_paras[3]
    Cascade=extra_paras[4]

    # Getting the net acceleration required in the global frame
    req_state_deri=guidance(t,state,extra)-gravity(t,state)
    anet=np.array([req_state_deri.vx,req_state_deri.vy,req_state_deri.vz])

    if t<0.01:
        print(f"t={t} anet=[{anet[0]}, {anet[1]}, {anet[2]}]")


    # Converting to body frame
    Anet=cons.global_to_body_matrix(state.phi,state.theta,state.psi)@anet

    # Calculate magnitude of the required acceleration
    anet_mag = np.linalg.norm(Anet)

    if anet_mag > 1e-6: # Prevent divide by zero
        anet_dir = Anet / anet_mag
    else:
        anet_dir = np.array([0.0, 0.0, 0.0])

    # Getting the error vector
    error_vec = np.array([-Anet[1], Anet[0], 0.0])

    # Getting required torque based on the error vector
    Torque=Cascade.run_cascade(error_vec,state,dt)

    # Getting the required RPMs based on the required torque and thrust
    Thrust=anet_mag*cons.mass
    RPMs=get_rpm(Thrust,Torque,motors)

    return RPMs

File: src/controller.cpp
#include "controller.h"
#include "guidance.h"
#include "gravity.h"
#include "environment.h"
#include "transformations.h"
#include "mixer.h"
using namespace std;

string phase = "IDLE";

// This is the controller responsible to calculate the required RPMs at any point of time
array<double,4> controllerOutput(double t, const State& state, const targetParas& target,const MotorSet& motors,Cascade& cascadeController){

    // Updating the mission phase based on the current state and target state
    updateMissionPhase(state,target.targetState);

    // Getting the net acceleration required in the global frame
    State reqStateDeri = guidance(phase,t,state,target)-gravity()-drag(t,state,rhoSurface,CdTrue,AreaTrue);
    Vector3 anet(reqStateDeri.vx,reqStateDeri.vy,reqStateDeri.vz);

    // Converting to bdoy frame
    Vector3 Anet = globalToBody(state.phi,state.theta,state.psi)*anet;

    // Calculate magnitude of the required acceleration
    double AnetMag = Anet.magnitude();

    // Getting the error vector
    Vector3 errorVec(-Anet[1],Anet[0],0.0);

    // Getting the required torque based on the error vector
    Vector3 Torque = cascadeController.runCascade(errorVec,state,target.dt);

    // Returning the required RPMs based on the required torque and thrust
    return getRPM(AnetMag*mass,Torque,motors);
}

// This function updates the mission phase based on the current state and target state
void updateMissionPhase(const State& state, const targetParas& target){
    
    // Getting the variables
    Vector3 rT=target.targetState;
    Vector3 rM(state.x,state.y,state.z);
    Vector3 rTM = rT-rM;
    Vector3 vM(state.vx,state.vy,state.vz);

    double cosTheta = (vM*rTM)/max(vM.magnitude()*rTM.magnitude(),denMin);

    // If the drone is close to the ground and in IDLE or rise phase, swtich to rise phase
    if(state.z<0.5 && (phase=="IDLE" || phase=="rise"))
        phase="rise";
    
    // If the drone is in rise or pure pursuit phase
    else if (phase=="rise" || phase=="purePursuit"){
        // If the drone is well aligned witht the target then check its speed else stay in pure pursuit phase
        if(cosTheta>0.5){
            //  If the drone is moving fast enough, switch to propNavPhase, else swtich to pure pursuit phase
            if(vM.magnitude()>0.5)
                phase="propNav";
            else
                phase="purePursuit";
        }
        else
            phase="purePursuit";
    }
    
    // If the drone is in propNav phase then stay in it
    else
        phase="propNav";
}